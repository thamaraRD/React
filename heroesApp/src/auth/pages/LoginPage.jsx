import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {

  const { login } = useContext(AuthContext);
  const navigate = useNavigate();

  const handleLogin = () => {

    //Para navegar al Ãºltimo paso antes de cerrar sesion
    const lastPath = localStorage.getItem('lasthPath') || '/';

    login( 'Fulanito' )
    //Se llama desde el navigate
    navigate(lastPath, {
      replace: true
    })
  }

  return (
    <div className="container mt-5">
    <h1>Login</h1>
    <hr />

    <button
    className="btn btn-outline-info"
    onClick={ handleLogin }
    >
    Login
    </button>
    </div>
  )
}
