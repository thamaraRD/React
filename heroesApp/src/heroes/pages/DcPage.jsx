import { HeroeList } from '../components/HeroeList';

export const DcPage = () => {
  return (
    <>
    <h1>Dc Comics</h1>
    <hr />

    <HeroeList publisher='DC Comics' />
    </>

  )
}
