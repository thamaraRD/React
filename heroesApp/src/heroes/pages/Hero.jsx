import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers/getHeroById";

export const Hero = () => {

  const { id } = useParams();
  const navigate = useNavigate();

  const hero = useMemo( ()=> getHeroById(id), [ id ] );

  const onNavigateBack = () =>{
  navigate("/");
  }

 if (!hero){
  return < Navigate to="/marvel" />
 }

 const imageUrl = `/assets/heroes/${id}.jpg`;

  return (
    <div className="row mt-5 animate__animated animate__fadeInTopRight">
    <div className="col-4">
      <img 
      src={imageUrl} 
      alt={hero.superhero}
      className="img-thumbnail"
      />
    </div>

    <div className="col-8">
      <h3>{hero.superhero}</h3>
      <ul className="list-group list-group-flush">
        <li className="list-group-item"><b>Alter ego: </b>{hero.alter_ego}</li>
        <li className="list-group-item"><b>Publisher: </b>{hero.publisher}</li>
        <li className="list-group-item"><b>First appearance: </b>{hero.first_appearance}</li>
      </ul>

      <h5 className="mt-5">Characters</h5>
      <p>{hero.characters}</p>

      <button 
      className="btn btn-primary"
      onClick={onNavigateBack}
      >
        Back
      </button>
    </div>
    </div>
  )
}
