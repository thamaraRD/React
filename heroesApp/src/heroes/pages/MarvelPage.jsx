import { HeroeList } from "../components/HeroeList"

export const MarvelPage = () => {
  return (
    <>
    <h1>Marvel Comics</h1>
    <hr />

    <HeroeList publisher='Marvel Comics' />

    </>
  )
}
