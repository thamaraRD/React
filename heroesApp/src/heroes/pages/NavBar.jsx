import { Link, NavLink, useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/context/AuthContext";
import { useContext } from "react";

export const NavBar = () => {

  const { user, logout } = useContext(AuthContext);

  const navigate = useNavigate();

  const handleLogout = () => {
    logout();
    navigate('/login', {
      replace: true
    });
  };

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <Link className="navbar-brand ms-2" to="/">
        Asociaciones
      </Link>

      <div className="navbar-collapse">
        <div className="navbar-nav">
          <NavLink className="nav-item nav-link" to="/marvel">
            Marvel
          </NavLink>

          <NavLink className="nav-item nav-link" to="/dc">
            DC
          </NavLink>

          <NavLink className="nav-item nav-link" to="/search">
            Buscar
          </NavLink>
        </div>
      </div>

      <div className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end">
        <ul className="navbar-nav ml-auto">
      <span className="nav-item nav-link text-info">
        {user?.name}
      </span>
      <button
      className="nav-item nav-link btn btn-outline-info me-2"
      onClick={handleLogout}
      >
        Logout
      </button>
        </ul>
      </div>
    </nav>
  );
};
