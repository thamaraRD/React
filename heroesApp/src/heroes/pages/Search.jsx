import {useLocation, useNavigate} from "react-router-dom";
import queryString from 'query-string';
import { useForm } from "../../hooks/useForm";
import { HeroeCard } from "../components/HeroeCard";
import { getHeroesByname } from "../helpers/getHeroesByName";

export const Search = () => {

const navigate = useNavigate();
const location = useLocation();

const { q = '' } = queryString.parse(location.search);
const heroes = getHeroesByname(q);


  const { searchText, onInputChange } = useForm({
    searchText: q
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();
    // if(searchText.trim().length <= 1) return;
    
    navigate(`?q=${searchText}`)
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
    <div className="row">
      <div className="col-5">
        <h4>Búsqueda</h4>
        <hr />
        <form onSubmit={ onSearchSubmit } aria-label="form">
          <input
            type="text"
            placeholder="Ingresa el heroe..."
            className="form-control p-3"
            name="searchText"
            autoComplete="off"
            value={searchText}
            onChange={onInputChange}
          />
        <button className="mt-3 btn btn-success">Buscar</button>
        </form>
      </div>
      <div className="col-7">
      <h4>Resultado</h4>
      <hr />
    {
      ( q === '' ) ? <div className="alert alert-success">Busca tu Heroe</div>
      :
      ( heroes.length === 0 ) && <div className="alert alert-danger">
        No se encontraron resultados de búsqueda para <b>{q}</b>
        </div>
    }

      {
        heroes.map( hero => ( 
        <HeroeCard key={hero.id} {...hero}/> 
        ) )
      }

      </div>
    </div>
    </>
  );
};
