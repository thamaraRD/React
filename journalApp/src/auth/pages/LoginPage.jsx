import { useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {Link as RouterLink} from 'react-router-dom'
import { Google } from "@mui/icons-material";
import { Button, Grid, Link, TextField, Typography } from "@mui/material";
import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks/useForm';
import { checkingAutentication, startGoogleSignIn } from '../../store/auth/thunk';

export const LoginPage = () => {

  const { status } = useSelector( state => state.auth );

  const dispatch = useDispatch();

  const { email, password, onInputChange, formState } = useForm({
    email: 'thata@gmail.com',
    password: '123456'
  });

  const isAuthenticating = useMemo( () => status === 'checking', [status] );

  const onSubmit = ( e ) => {
    e.preventDefault();

    console.log({ email, password });

    dispatch(checkingAutentication());
  }

  const onGoogleSignIn = () => {
    console.log('con google');
    dispatch(startGoogleSignIn());
  }

  return (
   <AuthLayout title='Login'>
        <form onSubmit={onSubmit} >
          <Grid container>
            <Grid item xs={12} sx={{mt: 2}}>
              <TextField
                label="Correo"
                type="email"
                placeholder="correo@dominio.com"
                fullWidth
                name='email'
                value={ email }
                onChange={ onInputChange }
              />
            </Grid>
            <Grid item xs={12} sx={{mt: 2}}>
              <TextField
                label="ContraseÃ±a"
                type="password"
                placeholder="******"
                fullWidth
                name='password'
                value={ password }
                onChange={ onInputChange }
              />
            </Grid>
            <Grid container spacing={2} sx={{mb: 2, mt: 1}}>
            <Grid item xs={12} md={6}>
            <Button
            disabled={isAuthenticating}
            type='submit'
            variant="contained"
            fullWidth
            >
              Login
            </Button>
            </Grid>
            <Grid item xs={12} md={6}>
            <Button
            disabled={isAuthenticating}
            variant="contained"
            fullWidth
            onClick={ onGoogleSignIn }
            >
              <Google />
              <Typography sx={{ml: 1}}>
                Google
              </Typography>
            </Button>
            </Grid>
            </Grid>
            <Grid container direction='row' justifyContent='end'>
              <Link color='inherit' to='/auth/register' component={RouterLink}>
             Crear una cuenta
              </Link>
            </Grid>
          </Grid>
        </form>
   </AuthLayout>

  );
};
