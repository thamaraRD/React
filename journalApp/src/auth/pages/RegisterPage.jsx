import {Link as RouterLink} from 'react-router-dom';
import { Button, Grid, Link, TextField, Typography } from "@mui/material";
import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks/useForm';

const formData = {
  email: 'rdthata@gmail.com',
  password: '123456',
  displayName: 'Thamara R'
};

const formValidations = {
    email: [ (value) => value.includes('@'), 'El correo debe tener una @' ],
    password: [ (value) => value.length >= 6, 'El password debe de tener más de 6 caracteres' ],
    displayName: [ (value) => value.length >= 1, 'El nombre es requerido' ]
}

export const RegisterPage = () => {

  const { formState, displayName ,email, password, onInputChange,
          isFormValid, displayNameValid, emailValid, passwordValid} = useForm(formData, formValidations);

  const onSubmit = ( event ) => {
    event.preventDefault();
    console.log(formState);
  }

  return (
   <AuthLayout title='Registro'>

        <form onSubmit={ onSubmit } >
          <Grid container>
            <Grid item xs={12} sx={{mt: 2}}>
              <TextField
                label="Nombre completo"
                type="text"
                placeholder="John Doe"
                fullWidth
                name='displayName'
                value={displayName}
                onChange={onInputChange}
                error={!displayNameValid}
                helperText={displayNameValid}
              />
            </Grid>
            <Grid item xs={12} sx={{mt: 2}}>
              <TextField
                label="Correo"
                type="email"
                placeholder="correo@dominio.com"
                fullWidth
                name='displayName'
                value={email}
                onChange={onInputChange}
              />
            </Grid>
            <Grid item xs={12} sx={{mt: 2}}>
              <TextField
                label="Contraseña"
                type="password"
                placeholder="******"
                fullWidth
              />
            </Grid>
            <Grid container spacing={2} sx={{mb: 2, mt: 1}}>
            <Grid item xs={12}>
            <Button
            type='submit'
            variant="contained"
            fullWidth
            >
              Crear Cuenta
            </Button>
            </Grid>
            </Grid>
            <Grid container direction='row' justifyContent='end'>
              <Typography sx={{mr: 1}}>¿Ya tienes cuenta?</Typography>
              <Link color='inherit' to='/auth/login' component={RouterLink}>
             ingresar
              </Link>
            </Grid>
          </Grid>
        </form>
   </AuthLayout>

  );
};
