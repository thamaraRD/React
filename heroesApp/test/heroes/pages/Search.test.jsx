import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Search } from "../../../src/heroes/pages/Search";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate

}));

describe('Pruebas en <Search />', () => { 

  beforeEach(() => jest.clearAllMocks());

  test('should de mostrarse correctamente con valores por defecto', () => { 

    const { container } = render(
      <MemoryRouter>
        <Search/>
      </MemoryRouter>
    )

   expect(container).toMatchSnapshot();

   })

   test('should de mostrar Batman y el input con el valor del queryString y la img ', () => { 

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search/>
      </MemoryRouter>
    )
      const input = screen.getByRole('textbox');
      expect( input.value ).toBe('batman');

      const img = screen.getByRole('img');
      expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
  

      // screen.debug();
    })


    test('should de llamar el navigate a la pantalla nueva', () => { 

      render(
        <MemoryRouter initialEntries={['/search']}>
          <Search/>
        </MemoryRouter>
      )

      const input = screen.getByRole('textbox');
      fireEvent.change( input, {target: { name: 'searchText', value: 'superman' }} )

      const form = screen.getByRole('form');
      fireEvent.submit(form);

      expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman')

     })

  })
