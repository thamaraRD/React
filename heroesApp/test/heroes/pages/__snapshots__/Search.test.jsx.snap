// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Search /> should de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        BÃºsqueda
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control p-3"
          name="searchText"
          placeholder="Ingresa el heroe..."
          type="text"
          value=""
        />
        <button
          class="mt-3 btn btn-success"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultado
      </h4>
      <hr />
      <div
        class="alert alert-success"
      >
        Busca tu Heroe
      </div>
    </div>
  </div>
</div>
`;
